### PR TITLE
Consul integration

### PR DESCRIPTION
**Pull request by falkerson**
Signed-off-by: Andriy Popovych <popovych.andrey@gmail.com>

### CODE REVIEW

**Comment by falkerson** [comment]
Partial fix for #6

**Comment by cloudxxx8** [comment]
@falkerson 
Will you continue to work on this PR?
If you are busy recently, may I take over this item?
Thanks

File: command.go [hunk 1]
```diff
@@ -37,15 +37,15 @@ func commandFunc(s *Service, w http.ResponseWriter, r *http.Request) {
 	d := dc.DeviceById(id)
 	if d == nil {
 		// TODO: standardize error message format (use of prefix)
-		msg := fmt.Sprintf("dev: %s not found; %s %s", id, r.Method, r.URL)
-		svc.lc.Error(msg)
+		msg := fmt.Sprintf("device: %s not found; %s %s", id, r.Method, r.URL)
+		s.lc.Error(msg)
```
[the remark this answers is not in the document]
Comment (falkerson) [reply command.go]: @tonyespy yes we can do it, but with current implementation seems we trap in dependency inversion problem https://en.wikipedia.org/wiki/Dependency_inversion_principle. A lot of functionality based on global Service object and it's hard to test.

File: command.go [hunk 2]
```diff
@@ -37,15 +37,15 @@ func commandFunc(s *Service, w http.ResponseWriter, r *http.Request) {
 	d := dc.DeviceById(id)
 	if d == nil {
 		// TODO: standardize error message format (use of prefix)
-		msg := fmt.Sprintf("dev: %s not found; %s %s", id, r.Method, r.URL)
-		svc.lc.Error(msg)
+		msg := fmt.Sprintf("device: %s not found; %s %s", id, r.Method, r.URL)
```
[the remark this answers is not in the document]
Comment (falkerson) [reply command.go]: yes, sorry but many tests are broken on my local env. Also I'm wonder how https://github.com/edgexfoundry/device-sdk-go/blob/master/command_test.go#L82 can pass if it expect different msg

File: commandresult_test.go [hunk 3]
```diff
@@ -62,7 +62,7 @@ func TestNewStringResult(t *testing.T) {
 	}
 
 	if result != cr.StringResult {
-		t.Errorf("NewStringResult: cr.StringResult: %s doesn't match result: %s")
+		t.Errorf("NewStringResult: cr.StringResult: %s doesn't match result: %s", cr.StringResult, result)
```
[the remark this answers is not in the document]
Comment (falkerson) [reply commandresult_test.go]: Already fixed in https://github.com/edgexfoundry/device-sdk-go/commit/0f8749cea2c79feded1791a0759736efe25b6263#diff-0ecc3d264d0217dc42b2c95981f27817R65

File: devices.go [hunk 4]
```diff
@@ -21,13 +21,13 @@ import (
 
 // deviceCache is a local cache of devices seeded from Core Metadata.
 type deviceCache struct {
-	devices     map[string]*models.Device
-	names       map[string]string
+	devices map[string]*models.Device
+	names   map[string]string
```
[the remark this answers is not in the document]
Comment (falkerson) [reply devices.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-1a60e1c4cc0c82b6f9d8eef9bcb80257R24

File: devices.go [hunk 5]
```diff
@@ -21,13 +21,13 @@ import (
...
 }
 
 var (
-	dcOnce  sync.Once
-	dc      *deviceCache
+	dcOnce sync.Once
+	dc     *deviceCache
```
[the remark this answers is not in the document]
Comment (falkerson) [reply devices.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-1a60e1c4cc0c82b6f9d8eef9bcb80257R29

File: examples/simple/cmd/main.go [hunk 6]
```diff
@@ -15,8 +15,8 @@ import (
 	"os/signal"
 	"syscall"
 
-	"github.com/edgexfoundry/device-sdk-go/examples/simple"
 	device "github.com/edgexfoundry/device-sdk-go"
+	"github.com/edgexfoundry/device-sdk-go/examples/simple"
```
[the remark this answers is not in the document]
Comment (falkerson) [reply examples/simple/cmd/main.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-33df89a0b0c78359d7a88eb3446803a6R19

File: examples/simple/simpledriver.go [hunk 7]
```diff
@@ -12,9 +12,9 @@ package simple
 import (
 	"fmt"
 
-	"github.com/edgexfoundry/edgex-go/pkg/models"
-	logger "github.com/edgexfoundry/edgex-go/pkg/clients/logging"
 	device "github.com/edgexfoundry/device-sdk-go"
+	logger "github.com/edgexfoundry/edgex-go/pkg/clients/logging"
+	"github.com/edgexfoundry/edgex-go/pkg/models"
```
[the remark this answers is not in the document]
Comment (falkerson) [reply examples/simple/simpledriver.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-fd1f38b5819842ad91e2ae05d650d3f1R16

File: profiles.go [hunk 8]
```diff
@@ -448,9 +447,9 @@ func (p *profileCache) createDescriptor(name string, devObj models.DeviceObject)
 	svc.lc.Debug(fmt.Sprintf("ps: createDescriptor: %v value: %v units: %s\n", name, value, units))
 
 	desc := &models.ValueDescriptor{
-		Name: name,
-		Min:  value.Minimum,
-		Max:  value.Maximum,
+		Name:         name,
+		Min:          value.Minimum,
+		Max:          value.Maximum,
```
[the remark this answers is not in the document]
Comment (falkerson) [reply profiles.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-ffe7b786b78992ada72fd65e595624a4R450

File: examples/simple/simpledriver.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (falkerson) [reply examples/simple/simpledriver.go]: Done in https://github.com/edgexfoundry/device-sdk-go/pull/35

File: command.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (falkerson) [reply command.go]: Done in https://github.com/edgexfoundry/device-sdk-go/pull/43

File: service.go [hunk 9]
```diff
@@ -22,10 +21,10 @@ import (
 	"time"
 
 	"github.com/edgexfoundry/edgex-go/pkg/clients/coredata"
+	logger "github.com/edgexfoundry/edgex-go/pkg/clients/logging"
```
[the remark this answers is not in the document]
Comment (falkerson) [reply service.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-30b92219c8fd27dd239b4dcd4bb4d838R24

File: service.go [hunk 10]
```diff
@@ -22,10 +21,10 @@ import (
...
 	"github.com/edgexfoundry/edgex-go/pkg/clients/metadata"
 	"github.com/edgexfoundry/edgex-go/pkg/clients/types"
 	"github.com/edgexfoundry/edgex-go/pkg/models"
-	logger "github.com/edgexfoundry/edgex-go/pkg/clients/logging"
```
[the remark this answers is not in the document]
Comment (falkerson) [reply service.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-30b92219c8fd27dd239b4dcd4bb4d838L28

File: service.go [hunk 11]
```diff
@@ -228,7 +227,17 @@ func (s *Service) Start(useRegistry bool, profile string, confDir string) (err e
...
+	if useRegistry {
+		consulMsg = "Loading configuration from Consul..."
+		err := connectToConsul(s.Name, s.c)
+		if err != nil {
+			return err
+		}
+	} else {
+		consulMsg = "Bypassing Consul configuration..."
+	}
+	fmt.Print(consulMsg)
```
[the remark this answers is not in the document]
Comment (falkerson) [reply service.go]: The question is next: should we Register/Load config with Consul before logger initialization or after? (For example some log related config info can be stored in Consul) In this case we do it before, so there is no logger and we use just print.

File: examples/simple/simpledriver.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (falkerson) [reply examples/simple/simpledriver.go]: Fixed in https://github.com/edgexfoundry/device-sdk-go/pull/35/files#diff-fd1f38b5819842ad91e2ae05d650d3f1R16

[the remark this answers is not in the document]
Comment (falkerson) [reply examples/simple/simpledriver.go]: Done in https://github.com/edgexfoundry/device-sdk-go/pull/35

File: utils_test.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
+package device
```
[the remark this answers is not in the document]
Comment (falkerson) [reply utils_test.go]: Done

File: utils.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
+package device
```
[the remark this answers is not in the document]
Comment (falkerson) [reply utils.go]: Done

File: utils.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+			HTTP:     config.Registry.CheckAddress,
+			Interval: config.Registry.CheckInterval,
+		},
+	})
+
+	return err
+}
+
+// TODO(apopovych) Store config data into Consul
+func storeConfigInConsul(consul *consulapi.Client, config *Config) error {
```
[the remark this answers is not in the document]
Comment (falkerson) [reply utils.go]: I thought we also want to store config in Consul. If not I'll just prepare patch with register.

[the remark this answers is not in the document]
Comment (falkerson) [reply utils.go]: Removed

File: config.go [hunk 15]
```diff
@@ -51,6 +44,16 @@ type service struct {
 	Host string
 	// Port is the HTTP port of a service.
 	Port int
+	// Health check url
+	CheckAddress string
+	// Health check interval
+	CheckInterval string
+	// ???
```
[the remark this answers is not in the document]
Comment (falkerson) [reply config.go]: removed - it's not needed for Consul

File: examples/simple/cmd/res/configuration.toml [hunk 16]
```diff
@@ -11,6 +11,9 @@ Timeout = 5000
 [Registry]
 Host = "localhost"
 Port = 8500
+CheckAddress = "http://localhost:49990/api/v1/ping"
+CheckInterval = "10s"
+ProfilesActive = "go"
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/cmd/res/configuration.toml]: he added

File: service.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply service.go]: We should start using loggerClient after loading configuration and initializing loggingCleint

**Comment by cloudxxx8** [comment]
@falkerson thanks for your contribution.  I would merge this PR first and continue do some minor change for this Consul integration tomorrow.

**Comment by cloudxxx8** [comment]
@tonyespy I would dismiss your PR, because @falkerson has canceled the unrelated commits and made the commits concisely.

File: command.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply command.go]: it is not in the PR now

File: .gitignore [hunk 17]
```diff
@@ -1,6 +1,6 @@
 *~
 *.log
-examples/simple/cmd/cmd
+examples/simple/cmd/simple-device
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply .gitignore]: it is not in this PR now

File: utils.go [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+		return err
+	}
+
+	// Register the Health Check
+	err = consul.Agent().CheckRegister(&consulapi.AgentCheckRegistration{
+		Name:      "Health Check: " + deviceServiceName,
+		Notes:     "Check the health of the API",
+		ServiceID: deviceServiceName,
+		AgentServiceCheck: consulapi.AgentServiceCheck{
+			HTTP:     config.Registry.CheckAddress,
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply utils.go]: it is dynamical now

File: examples/simple/cmd/res/configuration.toml [hunk 19]
```diff
@@ -11,6 +11,9 @@ Timeout = 5000
 [Registry]
 Host = "localhost"
 Port = 8500
+CheckAddress = "http://localhost:49990/api/v1/ping"
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/cmd/res/configuration.toml]: He modified as your suggestion

File: examples/simple/simpledriver.go [hunk 20]
```diff
@@ -56,6 +56,7 @@ func (s *SimpleDriver) HandleOperation(ro *models.ResourceOperation,
 	cr := &device.CommandResult{RO: ro, Type: device.Bool, BoolResult: true}
 
 	send <- cr
+	close(send)
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/simpledriver.go]: he canceled the change

[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/simpledriver.go]: he canceled the change

File: examples/simple/cmd/res/configuration.toml (hunk 19 above)
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/cmd/res/configuration.toml]: modified as your suggestion

[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/cmd/res/configuration.toml]: modified as your suggesstion

File: examples/simple/cmd/res/configuration.toml (hunk 16 above)
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply examples/simple/cmd/res/configuration.toml]: added

**Comment by cloudxxx8** [comment]
@tonyespy  falkerson has addressed your comment.  could you please approve this PR?

**Pull request merged**